client/zweb/renderer/index.js

```diff
@@ -103,6 +123,50 @@ class Renderer {
             Twig.exports.extendFunction("contract_get", async(target, contractName, method, params) => {
                 return await this.ctx.web3bridge.callContract(target, contractName, method, params);
             });
+            Twig.exports.extendFunction("contract_call", async(host, contractName, methodName, params) => {
+                return await this.ctx.web3bridge.sendContract(host.replace('.z', ''), contractName, methodName, params);
+            });
+            Twig.exports.extendFunction("retrieve_mail", async(owner, host, contractName, event) => {
+                const { privateKey } = this.ctx.wallet.config;
+                const events = await this.ctx.web3bridge.getPastEvents(host.replace('.z', ''), contractName, event, 0, 'latest');
+                const emails = await Promise.all(events.map(async (event)=> {
+                    if (owner === event.returnValues.to) {
+                        const encryptedSymmetricKey = JSON.parse(event.returnValues.encryptedSymmetricKey)
+                        const encryptedSymmetricObj = {}
+                        for (const k in encryptedSymmetricKey) {
+                            encryptedSymmetricObj[k] = Buffer.from(encryptedSymmetricKey[k], 'hex')
+                        }
+                        const encryptedMessage = await this.ctx.client.storage.readFile(event.returnValues.encryptedMessageHash, 'utf-8')
+                        const decryptedResult = await decryptCipherTextAndKey(host, Buffer.from(encryptedMessage, 'hex'), encryptedSymmetricObj, privateKey)
+                        return {
+                            messageid: event.returnValues.encryptedMessageHash,
+                            from: event.returnValues.from,
+                            message: decryptedResult.plaintext.toString(),
+                        }
+                    }
+                }))
+                return emails;
+            });
+            Twig.exports.extendFunction("decrypt_one_mail", async(host, owner, encryptedData) => {
+                const { privateKey } = this.ctx.wallet.config;
+                if (owner === encryptedData['1']) {
+                    const encryptedSymmetricKey = JSON.parse(encryptedData['3'])
```

Comment:
The body of this `if` block looks very similar to the email decryption happening in the `retrieve_mail` function above (lines 134-145). I think it is a good idea to move it to a separate function in order to tolerate possible changes in `decryptCipherTextAndKey` interface.